pretyping/evarconv.ml

```diff
@@ -440,16 +455,99 @@ and evar_eqappr_x ?(rhs_is_already_stuck = false) ts env evd pbty
+      else
+        UnifFailure (evd, NotSameHead)
+    in
+    let compare_heads evd = 
+      match EConstr.kind evd term, EConstr.kind evd term' with
+      | Const (c, u), Const (c', u') ->
+        check_strict ()
+      | Ind (ind, u), Ind (ind', u') ->
+        let check_subtyping_constraints () =
+          let nparamsaplied = Stack.args_size sk in
+          let nparamsaplied' = Stack.args_size sk' in
+          begin
+            let mind = Environ.lookup_mind (fst ind) env in
+            if mind.Declarations.mind_polymorphic then
+              begin
+                let num_param_arity =
+                  (* Context.Rel.length (mind.Declarations.mind_packets.(snd ind).Declarations.mind_arity_ctxt) *)
+                  mind.Declarations.mind_nparams + mind.Declarations.mind_packets.(snd ind).Declarations.mind_nrealargs
+                in
+                if not (num_param_arity = nparamsaplied && num_param_arity = nparamsaplied') then
+                  UnifFailure (evd, NotSameHead)
+                else
+                  begin
+                    let uinfind = mind.Declarations.mind_universes in
+                    let evd' = check_leq_inductives evd uinfind u u' in
+                    Success (check_leq_inductives evd' uinfind u' u)
+                  end
+              end
+            else
+              UnifFailure (evd, NotSameHead)
+          end
+        in
+        first_try_strict_check (Names.eq_ind ind ind') u u' check_subtyping_constraints
+      | Construct (cons, u), Construct (cons', u') ->
+        let check_subtyping_constraints () =
+          let ind, ind' = fst cons, fst cons' in
+          let j, j' = snd cons, snd cons' in
+          let nparamsaplied = Stack.args_size sk in
+          let nparamsaplied' = Stack.args_size sk' in
+          let mind = Environ.lookup_mind (fst ind) env in
+          if mind.Declarations.mind_polymorphic then
+            begin
+              let num_cnstr_args =
+                let nparamsctxt =
+                  (* Context.Rel.length mind.Declarations.mind_params_ctxt *)
+                  mind.Declarations.mind_nparams + mind.Declarations.mind_packets.(snd ind).Declarations.mind_nrealargs
+                in
+                nparamsctxt + mind.Declarations.mind_packets.(snd ind).Declarations.mind_consnrealargs.(j - 1)
+              in
+              if not (num_cnstr_args = nparamsaplied && num_cnstr_args = nparamsaplied') then
+                UnifFailure (evd, NotSameHead)
+              else
+                begin
+                  let uinfind = mind.Declarations.mind_universes in
+                  let evd' = check_leq_inductives evd uinfind u u' in
+                  Success (check_leq_inductives evd' uinfind u' u)
+                end
+            end
+          else
+            UnifFailure (evd, NotSameHead)
+        in
+        first_try_strict_check (Names.eq_constructor cons cons') u u' check_subtyping_constraints
+      | _, _ -> anomaly (Pp.str "")
```

Comment:
Shouldn't it rather be [assert false], i.e. unreachable pattern-matching branch?